File: src/main/java/org/eluder/jadcrypt/Jadcrypt.java
package org.eluder.jadcrypt;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class Jadcrypt {

    public static String encrypt(String plain, String password, String salt) throws Exception {
        return encrypt(plain, password, salt, Presets.DEFAULTS);
    }

    public static String encrypt(String plain, String password, String salt, Presets presets) throws Exception {
        byte[] key = CryptUtils.pbkdf2(password, salt, presets);
        byte[] iv = CryptUtils.md5(salt);

        return CryptUtils.toHex(encryptRaw(plain.getBytes(Presets.CHARSET), key, iv));
    }

    public static byte[] encryptRaw(byte[] plain, byte[] key, byte[] iv) throws Exception {
        SecretKey secret = new SecretKeySpec(key, "AES");
        IvParameterSpec ips = new IvParameterSpec(iv);

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, secret, ips);
        return cipher.doFinal(plain);
    }

    public static String decrypt(String encrypted, String password, String salt) throws Exception {
        return decrypt(encrypted, password, salt, Presets.DEFAULTS);
    }

    public static String decrypt(String encrypted, String password, String salt, Presets presets) throws Exception {
        byte[] key = CryptUtils.pbkdf2(password, salt, presets);
        byte[] iv = CryptUtils.md5(salt);

        return new String(decryptRaw(CryptUtils.fromHex(encrypted), key, iv), Presets.CHARSET);
    }

    public static byte[] decryptRaw(byte[] encrypted, byte[] key, byte[] iv) throws Exception {
        SecretKey secret = new SecretKeySpec(key, "AES");
        IvParameterSpec ips = new IvParameterSpec(iv);

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, secret, ips);
        return cipher.doFinal(encrypted);
    }

}

File: src/main/js/jadcrypt.js
import crypto from 'crypto';
import { charset, presets as options } from './presets';
import { pbkdf2, md5 } from './cryptutils';

export function encryptRaw(plain, key, iv, presets, callback) {
  if (typeof presets === 'function') {
    callback = presets;
    presets = options.defaults;
  }
  if (typeof callback !== 'function') {
    throw new Error('No callback provided to encryptRaw');
  }

  let encrypted = undefined;
  try {
    const cipher = crypto.createCipheriv(`aes-${presets.keyBits}-cbc`, key, iv);
    const data = cipher.update(plain);
    const remaining = cipher.final();
    encrypted = Buffer.concat([data, remaining], data.length + remaining.length);
  } catch (err) {
    callback(err);
  }
  if (encrypted !== undefined) {
    callback(null, encrypted);
  }
}

export function encrypt(plain, password, salt, presets, callback) {
  if (typeof presets === 'function') {
    callback = presets;
    presets = options.defaults;
  }
  if (typeof callback !== 'function') {
    throw new Error('No callback provided to encrypt');
  }
  
  pbkdf2(password, salt, presets, (err, key) => {
    if (err) {
      callback(err);
    } else {
      md5(salt, (err, iv) => {
        if (err) {
          callback(err);
        } else {
          encryptRaw(Buffer.from(plain, charset), key, iv, presets, (err, encrypted) => {
            if (err) {
              callback(err);
            } else {
              callback(null, encrypted.toString('hex'));
            }
          });
        }
      });
    }
  });
}

export function decryptRaw(encrypted, key, iv, presets, callback) {
  if (typeof presets === 'function') {
    callback = presets;
    presets = options.defaults;
  }
  if (typeof callback !== 'function') {
    throw new Error('No callback provided to decryptRaw');
  }
  
  let decrypted = undefined;
  try {
    const cipher = crypto.createDecipheriv(`aes-${presets.keyBits}-cbc`, key, iv);
    const data = cipher.update(encrypted);
    const remaining = cipher.final();
    decrypted = Buffer.concat([data, remaining], data.length + remaining.length);
  } catch (err) {
    callback(err);
  }
  if (decrypted !== undefined) {
    callback(null, decrypted);
  }
}

export function decrypt(encrypted, password, salt, presets, callback) {
  if (typeof presets === 'function') {
    callback = presets;
    presets = options.defaults;
  }
  if (typeof callback !== 'function') {
    throw new Error('No callback provided to decrypt');
  }
  
  pbkdf2(password, salt, presets, (err, key) => {
    if (err) {
      callback(err);
    } else {
      md5(salt, (err, iv) => {
        if (err) {
          callback(err);
        } else {
          decryptRaw(Buffer.from(encrypted, 'hex'), key, iv, presets, (err, decrypted) => {
            if (err) {
              callback(err);
            } else {
              callback(null, decrypted.toString(charset));
            }
          });
        }
      });
    }
  });
}
